internal/storage/dropbox/dropbox.go

```diff
@@ -0,0 +1,267 @@
+package dropbox
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/dropbox/dropbox-sdk-go-unofficial/v6/dropbox"
+	"github.com/dropbox/dropbox-sdk-go-unofficial/v6/dropbox/files"
+	"github.com/offen/docker-volume-backup/internal/storage"
+	"golang.org/x/oauth2"
+)
+
+type dropboxStorage struct {
+	*storage.StorageBackend
+	client           files.Client
+	concurrencyLevel int
+}
+
+// Config allows to configure a Dropbox storage backend.
+type Config struct {
+	Endpoint         string
+	RefreshToken     string
+	AppKey           string
+	AppSecret        string
+	RemotePath       string
+	ConcurrencyLevel int
+}
+
+// NewStorageBackend creates and initializes a new Dropbox storage backend.
+func NewStorageBackend(opts Config, logFunc storage.Log) (storage.Backend, error) {
+	tokenUrl, _ := url.JoinPath(opts.Endpoint, "oauth2/token")
+
+	conf := &oauth2.Config{
+		ClientID:     opts.AppKey,
+		ClientSecret: opts.AppSecret,
+		Endpoint: oauth2.Endpoint{
+			TokenURL: tokenUrl,
+		},
+	}
+
+	isCITest := opts.Endpoint != "https://api.dropbox.com/"
```

Comment:
I'm a bit torn on this. Is there any way of having the code without logic that determines whether it's running tests or not? Also, what happens id someone wants to use some future Dropbx alternative with the same API, but another URL? What keeps us from mocking the token dance?

Would it be possible to call a different OAuth2 server in the tests?

Reply:
I didn't like it either, but it realizes any expired tokens and needs a valid one. So maybe a OAuth2 mock service can be used here yes - it won't be validated by the Dropbox mock server anyway, just the Go OAuth2 package must be happy (so it has to be able to retrieve a fresh Access Token from a given Refresh Token).